File: leetcode/81-search-in-rotated-sorted-array-ii.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Boolean}
def search(nums, target)
  -1 != _search(nums, target, 0, nums.size - 1)
end


def _search(nums, target, b, e)
  if b < e
    mid = (e - b) / 2 + b
    if nums[mid] == target
      b = mid
    elsif nums[mid] < target and nums[mid] < nums[e] and nums[e] >= target
      b = _search(nums, target, mid+1, e)
    elsif nums[mid] > target and nums[b] < nums[mid] and nums[b] <= target
      b = _search(nums, target, b, mid-1)
    else
      b = _search(nums, target, b, mid-1)
      b = b == -1 ? _search(nums, target, mid+1, e) : b
    end
  end
  nums[b] == target ? b : -1
end

File: leetcode/81-search-in-rotated-sorted-array-ii.cpp
/*
Follow up for "Search in Rotated Sorted Array":
What if duplicates are allowed?

Would this affect the run-time complexity? How and why?

Write a function to determine if a given target is in the array.
*/

class Solution {
public:
    bool search(vector<int>& nums, int target) {
        return search(nums, 0, nums.size(), target) != -1;
    }
private:
    // [begin, end)
    int search(const vector<int>& nums, int begin, int end, int target) {
        if (end == begin) {
            return -1;
        }
        if (end - begin == 1) {
            return nums[begin] == target ? begin : -1;
        }
        if (nums[begin] < nums[end-1]) {
            // normal binary search
            int mid = (begin + end) / 2;
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] > target) {
                return search(nums, begin, mid, target);
            } else {
                return search(nums, mid+1, end, target);
            }
        } else if (target == nums[begin]) {
            return begin;
        } else {
            int mid = (begin + end) / 2;
            int left = search(nums, begin, mid, target);
            return left == -1 ? search(nums, mid, end, target) : left;
        }
    }
};
